src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/TagHelperBinder.cs

```diff
@@ -77,13 +78,48 @@ private static void ProcessDescriptors(
 
         foreach (var (key, value) in mapBuilder)
         {
-            map.Add(key, value.ToImmutableAndClear());
+            if (value is List<TagHelperDescriptor> builder)
+            {
+                map[key] = [.. builder];
+                ListPool<TagHelperDescriptor>.Default.Return(builder);
+            }
+            else
+            {
+                Debug.Assert(value is TagHelperDescriptor);
+                map[key] = [(TagHelperDescriptor)value];
+            }
         }
 
         tagNameToDescriptorsMap = new ReadOnlyDictionary<string, ImmutableArray<TagHelperDescriptor>>(map);
 
         // Build the catch all descriptors array.
         catchAllDescriptors = catchAllBuilder.ToImmutableAndClear();
+
+        static void AddToMapBuilder(Dictionary<string, object> mapBuilder, TagHelperDescriptor descriptor, string tagName)
+        {
+            if (!mapBuilder.TryGetValue(tagName, out var value))
+            {
+                // First descriptor for this tag name, just store it directly.
+                mapBuilder[tagName] = descriptor;
+            }
+            else
+            {
+                // If we have only seen a single descriptor for this tag name, upgrade to a list.
+                if (value is not List<TagHelperDescriptor> builder)
+                {
+                    Debug.Assert(value is TagHelperDescriptor);
+
+                    var existingDescriptor = (TagHelperDescriptor)value;
+                    builder = ListPool<TagHelperDescriptor>.Default.Get();
```

Comment:
I'm a little worried about using a pooled `List<TagHelperDescriptor>`. This probably works OK for OrchardCore where it's all legacy tag helpers. However, in something like MudBlazor where every .razor file represents a component, there are tons more `TagHelperDescriptors`. In those scenarios, we might end up starving the pool and creating new `List<TagHelperDescriptor>` instances anyway.

Reply:
No longer applicable in commit 2